addons/event/models/event_event.py

```diff
@@ -771,3 +774,107 @@ def _gc_mark_events_done(self):
         ])
         if ended_events:
             ended_events.action_set_done()
+
+    # Helpers for ESC/Label badge printing
+    def _get_event_timeframe_string(self):
+        self.ensure_one()
+        start_datetime = format_datetime(self.env, self.date_begin, self.date_tz, "short")
+        if self.is_one_day:
+            end_datetime = format_time(self.env, self.date_end, self.date_tz, "short")
+        else:
+            end_datetime = format_datetime(self.env, self.date_end, self.date_tz, "short")
+        return _("%(start_date)s to %(end_date)s", start_date=start_datetime, end_date=end_datetime)
+
+    def _print_event_template(self, layout: dict, flip=False):
+        self.ensure_one()
+        command = EscLabelCommand()
+
+        if self.badge_image:
+            command.print_image("BGFLIP" if flip else "BG", (0, layout["print_height"]) if flip else (0, 0))
+
+        print_centered_text(
+            layout,
+            text=self.name,
+            y_position=layout["event_name_y_pos"],
+            font_size=layout["event_name_font_size"],
+            command=command,
+            flip=flip
+        )
+
+        command.set_color(layout["secondary_text_color"], alpha=layout["secondary_text_alpha"])
+        print_centered_text(
+            layout,
+            text=self._get_event_timeframe_string(),
+            y_position=layout["date_y_pos"],
+            font_size=layout["details_font_size"],
+            command=command,
+            flip=flip
+        )
+
+        if self.address_id:
+            command.set_color(layout["secondary_text_color"], alpha=layout["secondary_text_alpha"])
+            print_centered_text(
+                layout,
+                text=self.address_id.name,
+                y_position=layout["address_y_pos"],
+                font_size=layout["details_font_size"],
+                command=command,
+                flip=flip
+            )
+
+        if self.company_id.logo:
+            logo_x_pos = (layout["print_width"] - layout["logo_width"]) / 2
+            logo_y_pos = layout["print_height"] * 2 - layout["logo_y_pos"] - layout["logo_height"] if flip else layout["logo_y_pos"]
+            command.print_image("LOGOFLIP" if flip else "LOGO", (logo_x_pos, logo_y_pos))
+
+        command.set_color(layout["secondary_text_color"], alpha=layout["secondary_text_alpha"])
+        print_centered_text(
+            layout,
+            text="Printed by Odoo Events App with Epson C4000e",
+            y_position=layout["custom_text_y_pos"],
+            font_size=layout["details_font_size"],
+            command=command,
+            flip=flip
+        )
+
+        return command
```

Comment:
Security-wise, to not escape the sandbox of safe_eval, it's better not to provide complex objects in model methods. I can explain you in DM.

It would be better to extract out of the models these methods returning a `EscLabelCommand` or a list/dict of them, and do the effort that model methods only return "simple" objects: string, booleans, integers, ...

From what I see, the only method regarding these "command" you absolutely need to be in the model is `_generate_esc_label_badges`, which is returning a simple string, which is the one called in the XML report.

Move all the rest of the methods returning a command out of the model. One possibility, as you already have a tool file for this, would be to put these methods in `addons/event/tools/esc_label_tools.py`

That being said, it's easier said then done.
- Ideally, you don't really want to pass browse records into tooling methods. So, instead of a record of attendee for instance, you would need to pass the actual data required (the attendee name, company name, ...)
- Moving things into a tool allows less customization, as you cannot override a tool as you can override an Odoo model/methods. But I guess at this point we will care about the ability to customize this later :p.

Reply:
Not sure if I was clear enough, but you will need to make a kind of API in your tools, taking a dict or list as parameter, in which you can pass the image, text, you want to put where in the badge.

Reply:
I think I understand, I've updated all the tooling methods so they now take dicts instead of the records themselves, and moved them outside of the models. Let me know if I've got anything wrong though